src/gen_rules.ml

```diff
@@ -348,8 +344,27 @@ let gen ~contexts
   in
   let+ contexts = Fiber.parallel_map contexts ~f:make_sctx in
   let map = String.Map.of_list_exn contexts in
+  let sctxs = String.Map.map map ~f:(fun (module M : Gen) -> M.sctx) in
+  let generators = (String.Map.map map ~f:(fun (module M : Gen) -> M.gen_rules)) in
+  let () =
+    Build_system.set_packages (fun path ->
+      let open Option.O in
+      Option.value ~default:[] (
+        let* ctx_name, _ = Path.extract_build_context path in
+        let* sctx = String.Map.find sctxs ctx_name in
+        Path.Map.find (Install_rules.packages sctx) path))
+  in
   Build_system.set_rule_generators
-    (String.Map.map map ~f:(fun (module M : Gen) -> M.gen_rules));
-  String.Map.iter map ~f:(fun (module M : Gen) -> M.init ());
-  String.Map.map map ~f:(fun (module M : Gen) -> M.sctx);
+    (function
+      | Install ctx ->
+        Option.map (String.Map.find sctxs ctx) ~f:(fun sctx ->
+          (fun ~dir _ ->
+             Install_rules.gen_rules sctx ~dir:(Path.as_in_build_dir_exn dir);
+             Build_system.These String.Set.empty))
+      | Context ctx ->
+        String.Map.find generators ctx);
 
+  String.Map.iter map ~f:(fun (module M : Gen) ->
+    Build_system.handle_add_rule_effects (fun () ->
```

Comment:
Once we port odoc to use `Rules`, we will not need this, right?

Reply:
Yeah, I think so. This changes somewhat in #2089 and I hope will go away at some point.